### PR TITLE
Markers - Add timestamps

### PR DESCRIPTION
**Pull request by Freddo3000**
**When merged this pull request will:**
- Add a checkbox when creating/editing markers to add a timestamp suffix.
- Requires a watch
- Move IDCs to macros.

image
image

Credit to @milaq for https://github.com/milaq/arma3_timemarkers from which this was inspired.

### CODE REVIEW

**Comment by BrettMayson** [comment]
Should instead be an option under direction, rather than tucked in the corner

File: addons/markers/functions/fnc_canTimestamp.sqf
Comment (BrettMayson) [inline addons/markers/functions/fnc_canTimestamp.sqf:18]:
```suggestion
params [["_unit", ACE_player]];
```

File: addons/markers/functions/fnc_initInsertMarker.sqf
Comment (BrettMayson) [inline addons/markers/functions/fnc_initInsertMarker.sqf:237]:
```suggestion
    if (GVAR(timestampEnabled) && {[ACE_player] call FUNC(canTimestamp)}) then {
```

**Comment by Freddo3000** [comment]
> 
> 
> Should instead be an option under direction, rather than tucked in the corner

I figured that considering it ties into the description, it should be close to it. I also wanted to keep the vertical space down.

File: addons/markers/functions/fnc_initInsertMarker.sqf
Comment (bux) [inline addons/markers/functions/fnc_initInsertMarker.sqf:246]: maybe provide a setting for time format and 24/12 h display

**Comment by BrettMayson** [comment]
> I figured that considering it ties into the description, it should be close to it. I also wanted to keep the vertical space down.

Vertical space is less of a concern than UX, maybe it could go under description then rather than at the bottom to keep it closer to related fields (Description)

File: addons/markers/functions/fnc_initInsertMarker.sqf
Comment (Freddo3000) [reply addons/markers/functions/fnc_initInsertMarker.sqf:246]: image

**Comment by Freddo3000** [comment]
image

File: addons/markers/functions/fnc_initInsertMarker.sqf
Comment (jokoho48) [inline addons/markers/functions/fnc_initInsertMarker.sqf:33]: why do you add those spaces?

File: addons/markers/stringtable.xml
Comment (jokoho48) [inline addons/markers/stringtable.xml:160]: why is here styling in the stringtable?

File: addons/markers/functions/fnc_initInsertMarker.sqf
Comment (Freddo3000) [reply addons/markers/functions/fnc_initInsertMarker.sqf:33]: I find it more readable, but I can revert it otherwise.

Comment (jokoho48) [reply addons/markers/functions/fnc_initInsertMarker.sqf:33]: I think it breaks the readability but that a maintainer should decide. I think we have it nowhere else in ace. I am also not sure if it is against coding style.

File: addons/markers/stringtable.xml
Comment (Freddo3000) [reply addons/markers/stringtable.xml:160]: Skips needing to format it in script

Comment (jokoho48) [reply addons/markers/stringtable.xml:160]: but may break if we have community translators that don't know what that is and maybe change something. I think styling should never be in a translation medium.

File: addons/markers/functions/fnc_initInsertMarker.sqf
Comment (veteran29) [reply addons/markers/functions/fnc_initInsertMarker.sqf:33]: We usually don't do that here.

**Comment by PabstMirror** [comment]
https://github.com/Freddo3000/ACE3/pull/1
I want to prevent this when editing markers


I tested an idea I had that uses a special asci char (that just looks like a space) to indicate that a timecode follows.
fnc_initInsertMarker:95
```
    if !((markerText GVAR(editingMarker)) isEqualTo "") then {
        // fill text input with text from marker which is being edited
        private _originalText = markerText GVAR(editingMarker);
        private _timeIndex = _originalText find ((toString [129]) + "["); // 129 just looks like a space
        if (_timeIndex > 0 ) then { _originalText = _originalText select [0,_timeIndex]; }; // stip off timecode
        _text ctrlSetText _originalText;
    };
```

fnc_onButtonClickConfirm:30
```
        "%1%2[%3%4]",
        ctrlText _description,
        toString [129],....
```

This means when you edit a marker it will automatically update the timecode instead of appending it.

File: addons/markers/functions/fnc_initInsertMarker.sqf
Comment (Freddo3000) [reply addons/markers/functions/fnc_initInsertMarker.sqf:33]: Though is it something that I will have to change for this to be approved?

Comment (veteran29) [reply addons/markers/functions/fnc_initInsertMarker.sqf:33]: Personally I think yes, do we have any guidelines on stuff like this? @jonpas ?

**Comment by Drofseh** [comment]
I could see it being useful to have a history sometimes `TEXT [1345] [1353] [1405]`.

Maybe a second button to add instead of replace?

File: addons/markers/functions/fnc_initInsertMarker.sqf
Comment (PabstMirror) [reply addons/markers/functions/fnc_initInsertMarker.sqf:33]: informal slack poll had mixed to negative results for this type of formatting
probably ok to leave in the PR, but it might get cleaned up in later refactoring

**Comment by Freddo3000** [comment]
> 
> 
> I could see it being useful to have a history sometimes `TEXT [1345] [1353] [1405]`.
> 
> Maybe a second button to add instead of replace?

At that point you might as well place a new marker, imo.

**Comment by Drofseh** [comment]
> > I could see it being useful to have a history sometimes `TEXT [1345] [1353] [1405]`.
> > Maybe a second button to add instead of replace?
> 
> At that point you might as well place a new marker, imo.

yah maybe...

**Review by veteran29 — APPROVED** [review]
LGTM

File: addons/markers/functions/fnc_initInsertMarker.sqf
Comment (veteran29) [reply addons/markers/functions/fnc_initInsertMarker.sqf:33]: Resolving the conversation then.

**Review by PabstMirror — APPROVED** [review]

**Review by jonpas — APPROVED** [review]

File: addons/markers/functions/fnc_onButtonClickConfirm.sqf
Comment (dedmen) [inline addons/markers/functions/fnc_onButtonClickConfirm.sqf:34]: This doesn't seem right?
description, space, space, time

you give 4 parameter but only use 3, and the space seems wrong

Comment (Freddo3000) [reply addons/markers/functions/fnc_onButtonClickConfirm.sqf:34]: Yeah looks like I made a typo, should probably be `"%1%2[%3%4]"`

Comment (commy2) [reply addons/markers/functions/fnc_onButtonClickConfirm.sqf:34]: Yes, am-pm is completely broken in this pull request. Even if this part was repaired, am-pm will not be shown for 12 and 24-hours, and even if that part is repaired, it will show the wrong one.

Should all be fixed here: https://github.com/acemod/ACE3/pull/8034/files#diff-4b395ce5b5e0cc52ea30d46cdc2cb9d5fea1eb22d6ca4cfe885b60ef68b1e45eR48

**Pull request merged**
